lessons/misc/intro-bash/lesson.md

```diff
@@ -0,0 +1,354 @@
+
+Before starting, let's change into our `scripts` folder. Note, we could run the scripts from aany folder by specifying the full path to the file, but for now we'll change over.
+
+```
+cd ../scripts
+```
+
+We'll be executing the python script `hello_world.py`. First, let's make sure that we have Python and that the computer knows where to find it (more on this in a bit). `which` is a helpful command for this as it tells us which *executable file* is being run when we type a command in the shell.
+
+```
+which python
+```
+
+If successful, the shell will print the full path to the executable used in performing the command. Incidentally, these are the files that are executed when your using your computer for everyday use; they're execution is simply hidden because you're often using a GUI.
+
+Let's go ahead and run our script
+
+```
+python hello_world.py
+```
+
+We can similarly execute bash scripts in the same way.
+
+```
+bash print-numbers.sh
+```
+
+R scripts can be run using the `Rscript <script-name>` command.
+
+# Environment variables
+
+There are special variables called *environment variables* which are used by bash and other programs to affect behaviour. Here, we'll only cover two of them. The first is one of the most important: `$PATH`:
+
+```
+echo $PATH
+```
+
+That's a little tough to read. Let's put the directories in our `$PATH` on separate lines by piping the output of `$PATH` to the **tr**anslate characters command, which allows us to quickly substitude the `:` in `$PATH` with newline (`\n`) characters.
+
+```
+echo "$PATH" | tr ':' '\n'
+```
+
+Notice the `$` symbol. This is how variables are accessed in bash. *Environment variables* are always capitalized. The `$PATH` variable sotres the directories you computer searches for programs, commands, and utilitis that have been installed. If you installed something and can't get it to work on the command-line, it's likely because the *executable file* for the program is not in the computers `$PATH`. When we used the `which` command earlier, these are the directories that were searched. If the command in not in `$PATH`, nothing will be printed.
+
+```
+which made-up-command
+```
+
+Another useful environment variable is `$PYTHONPATH`. This is the directories that Python will search for installed Python modules.
+
+```
+echo $PYTHONPATH | tr ':' '\n'
+```
+
+# Bash profile and aliases
+
+It turns out we can more finely tune our shell environment and add shortcuts and customizations to make our lives easier. This is done using `.bash_profile` or `.bashrc` configuration files. You can read about the differences between these [here](https://apple.stackexchange.com/questions/51036/what-is-the-difference-between-bash-profile-and-bashrc). These are shell configuration files that are executed every time you start shell and are executed just prior to the bash prompt appearing in the shell. Let's first see whether you have a configuration file.
+
+```
+cat ~/.bashrc
```

Comment:
It's maybe worth mentioning that these live in your home directory and nowhere else - a `.bash_profile` in a non-home directory will not be automatically run - ie make these files in your home dir and keep them there